Template tag parser, template sample cmd, configure welcome change, ask helper

scragly opened the pull request:
I've added a parsing function for the message template setup used for custom welcome messages.
The parser uses regex to match and replace the template tags that match and exist as valid objects.

In order to assist people sampling custom welcome messages in their own guild first, as well as to let groups of people easily collaborate together to discuss, sample and decide on a custom welcome message, I've added a command called `template` that accepts a sample message using the template tags and gives a parsed output of the sample.

Additionally, since there was a need to setup twice a wait_for_reaction, I just make a helper function called ask that accepts the following args: message (str or embed), destination (channel or user), user_id, and an optional keyword arg called react_list that lets you override the default tick/cross reaction options.
This should help minimise repetitive wait_for_reaction code a little bit, but since it's outside of the scope of this feature branch, I have not actually applied this new function to any other reaction confirmation messages outside of the ones created in these edits.

FoglyOgly commented:
Sorry if I'm just not seeing it, but will the bot still send warnings if it fails to find Discord objects?

scragly commented:
No; I couldnt think of a quick clean way to track that now that its being matched enmass within a contained check function. Instead itll output the nonmatched string as it was written by the user and itll be obvious to them its not converted when theyre shown the preview after sending the message. 

Its possible my brain just fried out though and i can have a look after a sleep if theres a way to return a list of non-matched tags.

FoglyOgly commented:
In do_template, could you have an empty list and append any non-matched tags, then at the end, if that list is non-empty, return an error saying "Hey, here's all the stuff I couldn't match"? That's just my first thought on how that might be possible.

scragly commented:
`do_template` now outputs `(msg: str, errors: list)`

A simple `if errors` can check if the list is not empty, and it'll contain the original tags as typed by the user.

I've added in the cases for errors in configure, of which it will preview the message, show the list of non-matches as a warning, and ask the user to check the data and try again.

In `!template` it'll just append the errors at the bottom of the message.

Both `!template` and `!configure` will show the preview as embeds if `[msg]`. This embed preview happens even if there's errors (with the Warning not_found list added as a field)

I've added a warning in the Welcome Message explanation text that roles will always show as `@deleted-role` in DMs. The roles will show as their original string and will be in the not_found list output if it's not found in the server, and won't display any issues if it was found successfully so it shouldn't cause tooooo much confusion. If they want to double check though, they're able to view it within the guild with the `!template` command as a test just to make sure or ease their minds.

This should now be fine for squash merging.
